aarch64_extension.cpp

```diff
@@ -586,6 +734,135 @@ class AArch64ArchitectureExtension : public ArchitectureHook {
+      return false;
+    }
+
+    if (-1 == Rn_index) {
+      /* INS (general) */
+      if (Rd_elem_size != Rn_size) {
+        LogError("%#lx: Operand size mismatch: Vd.Ts size %zu, Rn size %zu",
+                 instr->address, Rd_elem_size, Rn_size);
+        return false;
+      }
+
+      il.AddInstruction(
+          il.SetRegister(Rn_size, subRd, il.Register(Rn_size, Rn)));
+    } else {
+      /* INS (element) */
+
+      if (ARM64_VAS_INVALID == detail->operands[1].vas) {
+        LogError("%#lx: Vector operand 1 missing arrangement specifier",
+                 instr->address);
+        return false;
+      }
+
+      size_t Rn_elem_size = vectorElementSize(detail->operands[1].vas);
+      if (Rd_elem_size != Rn_elem_size) {
+        LogError("%#lx: Vector operand element size mismatch: %zu != %zu",
+                 instr->address, Rd_elem_size, Rn_elem_size);
+        return false;
+      }
+
+      uint32_t subRn = getVectorSubregister(Rn, Rn_size, Rd_index);
+      if (0 == subRn) {
+        LogError("%#lx: Invalid vector element access in operand 1",
+                 instr->address);
+        return false;
+      }
+
+      il.AddInstruction(il.SetRegister(Rd_elem_size, subRd,
+                                       il.Register(Rn_elem_size, subRn)));
+    }
+
+    return true;
+  }
+
+  bool LiftSTR(cs_insn* instr, LowLevelILFunction& il) {
+    cs_arm64* detail = &(instr->detail->arm64);
+
+    if (detail->op_count != 2 && detail->op_count != 3) {
+      return false;
+    }
+
+    if (detail->operands[0].type != ARM64_OP_REG ||
+        detail->operands[1].type != ARM64_OP_MEM) {
+      return false;
+    }
+
+    uint32_t Rt = this->m_base->GetRegisterByName(
+        cs_reg_name(disassembler.Get(), detail->operands[0].reg));
+
+    size_t Rt_size = this->m_base->GetRegisterInfo(Rt).size;
+
+    ExprId address = liftMemOperand(il, detail->operands[1]);
+    ExprId value;
+
+    if (ARM64_REG_WZR == detail->operands[0].reg ||
+        ARM64_REG_XZR == detail->operands[0].reg) {
+      /* The zero register doesn't count */
+      value = il.Const(1, 0);
+    } else {
+      value = il.Register(Rt_size, Rt);
+    }
+
+    il.AddInstruction(il.Store(Rt_size, address, value));
+
+    if (detail->writeback) {
+      /* Lift write-back to base register (Rn) */
+      uint32_t Rn = this->m_base->GetRegisterByName(
+          cs_reg_name(disassembler.Get(), detail->operands[1].mem.base));
+      size_t Rn_size = this->m_base->GetRegisterInfo(Rn).size;
+
+      /* Lift address expr again for writeback */
+      address = liftMemOperand(il, detail->operands[1]);
+
+      if (3 == detail->op_count) {
+        /* Add post-index */
+        if (ARM64_OP_IMM != detail->operands[2].type) {
+          LogError("%#lx: Expected immediate or nothing for operand 2",
+                   instr->address);
+          return false;
+        }
+
+        address =
+            il.Add(Rn_size, address, il.Const(4, detail->operands[2].imm));
```

Comment:
Will this immediate always be of size 4?

Reply:
Added a range check above there, and changed to 2 bytes (technically 9 bits).